src/cache_cf.cc

```diff
@@ -1014,13 +1023,60 @@ parse_obsolete(const char *name)
     }
 }
 
-/* Parse a time specification from the config file.  Store the
- * result in 'tptr', after converting it to 'units' */
-static void
-parseTimeLine(time_msec_t * tptr, const char *units,  bool allowMsec,  bool expectMoreArguments = false)
+/// \returns the number of samples specified by 'unit' in unitName
+template <class T>
+T
+parseTimeUnits(const char *unitName, const TimeUnits unit)
 {
-    time_msec_t u = parseTimeUnits(units, allowMsec);
-    if (u == 0) {
+    if (unit >= TimeUnits::nanoSeconds && !strncasecmp(unitName, T_NANOSECOND_STR, strlen(T_NANOSECOND_STR)))
+        return unit/TimeUnits::nanoSeconds;
+
+    if (unit >= TimeUnits::microSeconds && !strncasecmp(unitName, T_MICROSECOND_STR, strlen(T_MICROSECOND_STR)))
+        return unit/TimeUnits::microSeconds;
+
+    if (unit >= TimeUnits::milliSeconds && !strncasecmp(unitName, T_MILLISECOND_STR, strlen(T_MILLISECOND_STR)))
+        return unit/TimeUnits::milliSeconds;
+
+    if (!strncasecmp(unitName, T_SECOND_STR, strlen(T_SECOND_STR)))
+        return unit;
+
+    if (!strncasecmp(unitName, T_MINUTE_STR, strlen(T_MINUTE_STR)))
+        return 60 * unit;
+
+    if (!strncasecmp(unitName, T_HOUR_STR, strlen(T_HOUR_STR)))
+        return 3600 * unit;
+
+    if (!strncasecmp(unitName, T_DAY_STR, strlen(T_DAY_STR)))
+        return 86400 * unit;
+
+    if (!strncasecmp(unitName, T_WEEK_STR, strlen(T_WEEK_STR)))
+        return 86400 * 7 * unit;
+
+    if (!strncasecmp(unitName, T_FORTNIGHT_STR, strlen(T_FORTNIGHT_STR)))
+        return 86400 * 14 * unit;
+
+    if (!strncasecmp(unitName, T_MONTH_STR, strlen(T_MONTH_STR)))
+        return 86400 * 30 * unit;
+
+    if (!strncasecmp(unitName, T_YEAR_STR, strlen(T_YEAR_STR)))
+        return static_cast<T>(86400 * unit * 365.2522);
+
+    if (!strncasecmp(unitName, T_DECADE_STR, strlen(T_DECADE_STR)))
+        return static_cast<T>(86400 * unit * 365.2522 * 10);
```

Comment:
Probably this method should check for integer overflows (as parseTimeLine() already checks for them). For now, I do not see an elegant way how to do this.